### PR TITLE
Make rocksdb work with gcc 7 (v5.4.6)

### PR DESCRIPTION
**Pull request by derekdreery**
Note that this upgrade causes some tests to fail in ffi, but I lack the skills to fix them.

### CODE REVIEW

**Comment by spacejam** [comment]
There are a few new options added to rocksdb's c API, which the test is failing on when they are not enabled. We don't have those new options in this crate's FFI code yet. The easiest path to getting this to build is to add them, and remove ones that no longer exist.

new:

rocksdb_options_set_allow_concurrent_memtable_write

removed:

rocksdb_options_set_allow_os_buffer (also relied on in the code)

there may be others, but this is where the compilation is currently getting hung up.  I found these by looking at the error message in the failed build:
```
thread 'ffi' panicked at 'prefix: Invalid argument: Memtable doesn't concurrent writes (allow_concurrent_memtable_write)', librocksdb-sys/tests/ffi.rs:1035
```
That message told me that there's an option that is disabled, called `allow_concurrent_memtable_write` which must be turned on for the usage in tests/ffi.rs:1035 to succeed. This is set to enabled by default in some cases, but disabled in others in the rocksdb source, so we need to figure out if there's some safety thing we need to look out for.

Then this message:
```
failures:

---- src/db_options.rs - db_options::Options::set_allow_os_buffer (line 375) stdout ----

	error: linking with `cc` failed: exit code: 1
...
```

told me that set_allow_os_buffer is no longer found, and looking in the rocksdb C API source, it indeed is no longer present:
https://github.com/facebook/rocksdb/blob/master/include/rocksdb/c.h

so we need to remove anything that touches it if we want to bump the rocksdb version.

Do you want to look into bumping these? My time is a bit limited, but I would welcome any version bump that doesn't break the build.

**Comment by derekdreery** [comment]
Thanks I'll have a look

**Comment by kaedroho** [comment]
I've just implemented a fix for that on my branch: https://github.com/kaedroho/rust-rocksdb/commit/31d71fac8c4894dbbd4c0c58c2d247a2c2c360ae

**Comment by derekdreery** [comment]
Awesome :)

**Comment by derekdreery** [comment]
I've cherry-picked those commits, and am just failing on one doctest. I believe the function it's looking for should be there, so have I got the wrong rocksdb commit?

File: src/db_options.rs [hunk 1]
```diff
@@ -399,39 +399,6 @@ impl Options {
...
-    /// # Example
-    ///
-    /// ```
-    /// use rocksdb::Options;
-    ///
-    /// let mut opts = Options::default();
-    /// opts.set_allow_os_buffer(false);
-    /// ```
-    #[deprecated(since="0.7.0", note="replaced with set_use_direct_reads/set_use_direct_writes methods")]
-    pub fn set_allow_os_buffer(&mut self, is_allow: bool) {
```
Comment (kaedroho) [inline src/db_options.rs]: Why did you remove this?

Comment (derekdreery) [inline src/db_options.rs]: I thought it had been removed from rocksdb - I'll check.

Comment (kaedroho) [inline src/db_options.rs]: It has been, but I reimplemented it on the new DirectIO methods with a deprecation warning

Comment (derekdreery) [inline src/db_options.rs]: I get the following output for thist doctest

```
---- src/db_options.rs - db_options::Options::set_allow_os_buffer (line 423) stdout ----
	warning: use of deprecated item: replaced with set_use_direct_reads/set_use_direct_writes methods
 --> <anon>:6:6
  |
6 | opts.set_allow_os_buffer(false);
  |      ^^^^^^^^^^^^^^^^^^^
  |
  = note: #[warn(deprecated)] on by default

error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/home/rdodd/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/tmp/rustdoctest.yFOM2FA1OKGb/rust_out.0.o" "-o" "/tmp/rustdoctest.yFOM2FA1OKGb/rust_out" "/tmp/rustdoctest.yFOM2FA1OKGb/rust_out.crate.allocator.o" "-Wl,--gc-sections" "-pie" "-nodefaultlibs" "-L" "/home/rdodd/.cargo/target_dir/debug/deps" "-L" "/home/rdodd/.cargo/target_dir/debug/build/librocksdb-sys-ff15d401df6cb8a6/out" "-L" "/home/rdodd/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/home/rdodd/.cargo/target_dir/debug/deps/librocksdb-e312e09db9ca21d5.rlib" "/home/rdodd/.cargo/target_dir/debug/deps/liblibrocksdb_sys-ae3566a82a4c0e36.rlib" "/home/rdodd/.cargo/target_dir/debug/deps/liblibc-d2710bfb355676bc.rlib" "-L" "/home/rdodd/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bdynamic" "-l" "std-1d2a102a93f1dfa9" "-Wl,-Bstatic" "/home/rdodd/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-821b4ec553d97f1e.rlib" "-Wl,-Bdynamic" "-l" "stdc++" "-l" "util" "-l" "dl" "-l" "rt" "-l" "pthread" "-l" "pthread" "-l" "gcc_s" "-l" "c" "-l" "m" "-l" "rt" "-l" "pthread" "-l" "util"
  = note: /home/rdodd/.cargo/target_dir/debug/deps/librocksdb-e312e09db9ca21d5.rlib(rocksdb-e312e09db9ca21d5.0.o): In function `rocksdb::db_options::{{impl}}::set_use_direct_writes':
          /home/rdodd/devel/rust-rocksdb/src/db_options.rs:398: undefined reference to `rocksdb_options_set_use_direct_writes'
          collect2: error: ld returned 1 exit status

```

Comment (derekdreery) [inline src/db_options.rs]: Aah got you, the merger of the cherry-pick must have destroyed it - I'll look at your commit.

Comment (kaedroho) [inline src/db_options.rs]: Ahh, looks like there's two issues there:

```
---- src/db_options.rs - db_options::Options::set_allow_os_buffer (line 423) stdout ----
	warning: use of deprecated item: replaced with set_use_direct_reads/set_use_direct_writes methods
 --> <anon>:6:6
  |
6 | opts.set_allow_os_buffer(false);
  |      ^^^^^^^^^^^^^^^^^^^
  |
  = note: #[warn(deprecated)] on by default
```
This is a warning because I deprecated that method but forgot to add ``#[allow(deprecated)]`` to its doctest.

```
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/home/rdodd/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/tmp/rustdoctest.yFOM2FA1OKGb/rust_out.0.o" "-o" "/tmp/rustdoctest.yFOM2FA1OKGb/rust_out" "/tmp/rustdoctest.yFOM2FA1OKGb/rust_out.crate.allocator.o" "-Wl,--gc-sections" "-pie" "-nodefaultlibs" "-L" "/home/rdodd/.cargo/target_dir/debug/deps" "-L" "/home/rdodd/.cargo/target_dir/debug/build/librocksdb-sys-ff15d401df6cb8a6/out" "-L" "/home/rdodd/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/home/rdodd/.cargo/target_dir/debug/deps/librocksdb-e312e09db9ca21d5.rlib" "/home/rdodd/.cargo/target_dir/debug/deps/liblibrocksdb_sys-ae3566a82a4c0e36.rlib" "/home/rdodd/.cargo/target_dir/debug/deps/liblibc-d2710bfb355676bc.rlib" "-L" "/home/rdodd/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bdynamic" "-l" "std-1d2a102a93f1dfa9" "-Wl,-Bstatic" "/home/rdodd/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-821b4ec553d97f1e.rlib" "-Wl,-Bdynamic" "-l" "stdc++" "-l" "util" "-l" "dl" "-l" "rt" "-l" "pthread" "-l" "pthread" "-l" "gcc_s" "-l" "c" "-l" "m" "-l" "rt" "-l" "pthread" "-l" "util"
  = note: /home/rdodd/.cargo/target_dir/debug/deps/librocksdb-e312e09db9ca21d5.rlib(rocksdb-e312e09db9ca21d5.0.o): In function `rocksdb::db_options::{{impl}}::set_use_direct_writes':
          /home/rdodd/devel/rust-rocksdb/src/db_options.rs:398: undefined reference to `rocksdb_options_set_use_direct_writes'
          collect2: error: ld returned 1 exit status
```

This is a separate issue because RocksDB has removed the ``rocksdb_options_set_use_direct_writes`` function

Comment (derekdreery) [inline src/db_options.rs]: Is there an alternative? Should I just remove this function?

Comment (kaedroho) [inline src/db_options.rs]: Looks like we should just rename it in ``our``` DBOptions. We would need it to implement backwards compatibility for the old ``set_allow_os_buffer`` method

File: src/db_options.rs [hunk 2]
```diff
@@ -388,17 +388,79 @@ impl Options {
...
     ///
     /// ```
+    /// #[allow(deprecated)]
     /// use rocksdb::Options;
     ///
     /// let mut opts = Options::default();
     /// opts.set_use_direct_writes(true);
     /// ```
+    #[deprecated(note = "replaced by `Options::set_use_direct_io_for_flush_and_compaction`")]
     pub fn set_use_direct_writes(&mut self, enabled: bool) {
```
Comment (kaedroho) [inline src/db_options.rs]: I think we can delete this method, as it was introduced in this PR anyway

Comment (derekdreery) [inline src/db_options.rs]: cool

File: src/db_options.rs [hunk 3]
```diff
@@ -388,17 +388,79 @@ impl Options {
...
         unsafe {
-            ffi::rocksdb_options_set_use_direct_writes(self.inner, enabled as c_uchar);
+            ffi::rocksdb_options_set_use_direct_io_for_flush_and_compaction(self.inner,
+                                                                            enabled as c_uchar);
         }
     }
 
+    /// Enable direct I/O mode for writing
```
Comment (kaedroho) [inline src/db_options.rs]: This needs to be reworded:

```
    /// Enable direct I/O mode for flush and compaction
    ///
    /// Files will be opened in "direct I/O" mode which means that data read and
    /// written to the disk will not be cached or buffered for flushes and compactions.
    /// The hardware buffer of the devices may however still be used. Memory
    /// mapped files are not impacted by these parameters.
```

Comment (derekdreery) [inline src/db_options.rs]: Yes I agree - I'll see if there are any docs for the new method.

Comment (derekdreery) [inline src/db_options.rs]: Oh I see, I can just copy and paste what you say.

**Comment by kaedroho** [comment]
Looking good so far!

Looks like there's one more issue to solve:
```
thread 'ffi' panicked at 'prefix: Invalid argument: Memtable doesn't concurrent writes (allow_concurrent_memtable_write)', librocksdb-sys/tests/ffi.rs:1035
```

**Comment by derekdreery** [comment]
That's the original issue I think - I'll have a look at fixing it.

**Comment by kaedroho** [comment]
Nice work @derekdreery!

@spacejam I think this is ready for you to review now

File: librocksdb-sys/src/lib.rs [hunk 4]
```diff
@@ -633,7 +635,9 @@ extern "C" {
     pub fn rocksdb_options_set_purge_redundant_kvs_while_flush(opt: *mut rocksdb_options_t,
                                                                v: c_uchar);
 
-    pub fn rocksdb_options_set_allow_os_buffer(opt: *mut rocksdb_options_t, v: c_uchar);
+    pub fn rocksdb_options_set_use_direct_reads(opt: *mut rocksdb_options_t, v: c_uchar);
+    pub fn rocksdb_options_set_use_direct_io_for_flush_and_compaction(opt: *mut rocksdb_options_t,
+                                                                      v: c_uchar);
```
Comment (spacejam) [inline librocksdb-sys/src/lib.rs]: minor nit: please retain spacing between functions

**Comment by spacejam** [comment]
published version 0.7.0

thanks for the help!
